src/Microsoft.IdentityModel.Tokens/MaxCapacityRefreshCache.cs

```diff
@@ -0,0 +1,137 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+//------------------------------------------------------------------------------
+
+namespace Microsoft.IdentityModel.Tokens
+{
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+
+    /// <summary>
+    /// A cache that clears itself upon reaching max capacity.
+    /// </summary>
+    /// <typeparam name="TKey">The type of cache key.</typeparam>
+    /// <typeparam name="TValue">The type of cached value.</typeparam>
+    internal class MaxCapacityRefreshCache
+        <TKey, TValue>
+        where TValue : class
+    {
+        /// <summary>
+        /// Gets or sets the cache.
+        /// </summary>
+        private ConcurrentDictionary<TKey, TValue> Cache { get; set; }
+
+        /// <summary>
+        /// Gets or sets the cache size limit.
+        /// </summary>
+        private long SizeLimit { get; set; }
+
+        /// <summary>
+        /// Gets or sets the lock for recycle operations.
+        /// </summary>
+        private object RecycleLock { get; set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MaxCapacityRefreshCache{TKey,TValue}" /> class.
+        /// </summary>
+        /// <param name="sizeLimit">The max number of items the cache is capable of holding.</param>
+        /// <param name="comparer">The string comparer.</param>
+        internal MaxCapacityRefreshCache(
+            long sizeLimit,
+            IEqualityComparer<TKey> comparer = null)
+        {
+            RecycleLock = new object();
+            SizeLimit = sizeLimit;
+            Cache = new ConcurrentDictionary<TKey, TValue>(comparer ?? EqualityComparer<TKey>.Default);
+        }
+
+        internal bool Contains(TKey key)
+        {
+            return Cache.ContainsKey(key);
+        }
+
+        /// <summary>
+        /// Gets the value corresponding to <paramref name="value"/> from the cache.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="value">The cached value.</param>
+        /// <remarks>This method is non-blocking and thread-safe.</remarks>
+        internal bool TryGetValue(TKey cacheKey, out TValue value)
+        {
+            RecycleCacheIfNeeded();
+            return Cache.TryGetValue(cacheKey, out value);           
+        }
+
+        /// <summary>
+        /// Sets the cache value.
+        /// </summary>
+        /// <param name="cacheKey">The cache key.</param>
+        /// <param name="value">The value.</param>
+        /// <returns><paramref name="value"/> if it was successfully set, null otherwise.</returns>
+        /// <remarks>This method is non-blocking and thread-safe.</remarks>
+        /// <remarks>This value will not be set if the cache already contains a key with the same name as <paramref name="cacheKey"/>.</remarks>
```

Comment:
This is a bit surprising behavior, why not let a set update the value in the cache?

Reply:
This is the default behavior of the ConcurrentDictionary class + is consistent with the current behavior of the InMemoryCryptoProviderCache.

I do agree that this is somewhat confusing behavior though; more than happy to change it if we agree this approach doesn't make sense.

Reply:
MemoryCache's Set method will update. It is the InMemoryCryptoProvider cache that we do a set if not exists.

The layering feels off here, Set should Set. Concurrent dictionary supports that with an indexer.

Pls let me know if I'm missing something

Reply:
Right, I agree that the MemoryCache's set method will update the cache.

I do think the [TryAdd()](https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentdictionary-2.tryadd?view=net-5.0) method from the ConcurrentDictionary prevents an already set value from being overridden though.

> `true` if the key/value pair was added to the ConcurrentDictionary<TKey,TValue> successfully; `false` if the key already exists.

I might be misunderstanding something as well.

Reply:
Seems like I confused myself by focusing too much on the ConcurrentDictionary APIs, haha. I've gone ahead and modified this code to make use of the indexer and to override the current value if it already exists.